Step5 - 로또(수동) PR

seogineer opened the pull request:
PR 올립니다.

seogineer committed:
refactor: merge comment에 있던 내용 반영

seogineer committed:
style: 상수명 변경

seogineer committed:
docs: 기능 목록 작성

seogineer committed:
refactor: 정적 팩토리 메소드 적용

seogineer committed:
docs: "구매한 로또 번호(List<LottoNumbers>)에 일급 컬렉션 적용" 내용 추가

seogineer committed:
refactor: 인스턴스 캐싱 적용

seogineer committed:
refactor: 구매한 로또 번호(List<LottoNumbers>)에 일급 컬렉션 적용

seogineer committed:
refactor: 구매한 로또 번호(List<LottoNumbers>)에 일급 컬렉션 적용

seogineer committed:
docs: 내용 추가

seogineer committed:
feat: 기능 추가

- 수동으로 구매할 로또 수 입력 기능 및 문구 추가
- 수동으로 구매할 로또 번호 입력 기능 및 문구 추가
- 수동으로 구매한 로또 수에 따라 자동으로 구매할 수 있는 개수 계산 기능 추가
-

seogineer committed:
feat: 기능 추가

- 수동으로 구매할 로또 수 입력 기능 및 문구 추가
- 수동으로 구매할 로또 번호 입력 기능 및 문구 추가
- 수동으로 구매한 로또 수에 따라 자동으로 구매할 수 있는 개수 계산 기능 추가
-

seogineer committed:
docs: Validate 메소드 이름 변경 추가

seogineer committed:
docs: Validate에 추가할 기능에 대한 내용 추가

seogineer committed:
style: 메소드 이름 변경

seogineer committed:
feat: 수동으로 번호 입력 받을때 유효성 검사 추가

seogineer committed:
feat: 수동으로 구매 개수 입력 받을때 유효성 검사 추가

seogineer committed:
docs: 불필요한 유효성 검사 항목 제거

seogineer committed:
refactor: 불필요한 유효성 검사 항목 제거 및 이

seogineer committed:
style: 더 정확한 변수 이름으로 수정

seogineer committed:
refactor: 사용자가 잘못된 값을 입력했을 Optional을 적용해 NullPointerException이 발생하지 않도록 한다.

seogineer committed:
refactor: LottoGame balance, totalMoney -> 일급 객체로 수정

seogineer committed:
refactor: manualPurchaseCount, autoPurchaseCount -> 일급 객체로 수정

seogineer committed:
docs: 일급 컬렉션 항목 수정

geunhwanlee commented on src/main/java/lotto/Balance.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.Objects;
+
+public class Balance {
```
`Balance`와 `TotalMoney`를 분리한 이유가 있을까요?🤔
`구입금액`과 `수동구매수량`을 입력받아 `자동구매수량`을 반환하는 역할을 하나의 객체가 수행할수 있지 않을까요?

geunhwanlee commented on src/main/java/lotto/InputView.java:
```diff
@@ -18,8 +20,9 @@ public class InputView {
         printInputPay();
-        String input = scanner.nextLine();
-        Validate.validateCostNull(input);
+        Optional<String> inputOptional = Optional.ofNullable(scanner.nextLine());
```
`Scanner.nextLine`이 `null`을 반환하는 케이스가 있을까요?🤔
그리고 null체크를 하는데 Optional로 감싸는것보단 if문으로 체크하는게 더 나을것같아요!

seogineer replied:
요구사항에 Optional로 null 체크하라는 요구사항이 있었는데 제가 Optional을 별로 사용해 본 경험이 없어서 요구사항에 맞춰서 하다 보니 사용 안해도 될 부분에 적용했습니다.

geunhwanlee commented on src/main/java/lotto/InputView.java:
```diff
@@ -29,28 +32,46 @@ private void printInputPay() {
+        printInputManualPurchase();
+        Optional<String> inputOptional = Optional.ofNullable(scanner.nextLine());
+        String input = inputOptional.orElse("0");
+        Validate.validatePurchasableCount(payMoney, Integer.parseInt(input));
```
View에서 유효성검사를 하는게 점점 복잡해지는것 같은데
적절한 객체에게 유효성검사를 위임하는건 어떨까요?🤔

geunhwanlee commented on src/main/java/lotto/LottoNumberGenerator.java:
```diff
@@ -1,12 +1,41 @@
-public interface LottoNumberGenerator {
-    List<LottoNumber> generateSixNumbers();
+    public List<Integer> manualGenerateNumbers() {
+        final Scanner scanner = new Scanner(System.in);
```
콘솔 입력을 받는건 view에서 해야할것 같아요!

geunhwanlee commented on src/main/java/lotto/LottoNumberGenerator.java:
```diff
@@ -1,12 +1,41 @@
+        return lottoNumberKeys.stream()
+                .limit(LOTTO_NUMBER_SIZE)
+                .collect(Collectors.toList());
+    }
```
`LottoNumbers`를 반환하는건 어떨까요?

geunhwanlee commented on src/main/java/lotto/PurchaseLottoNumbers.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class PurchaseLottoNumbers {
```
네이밍이 어렵네요..😭

> PurchaseLottoNumbers -> Lottos
> LottoNumbers -> Lotto

위 명칭이 더 이해하기 좋을것같은데 어떤가요?

geunhwanlee commented on src/main/java/lotto/ResultView.java:
```diff
@@ -16,17 +18,15 @@
+    void printResultPay(PurchaseCount autoPurchaseCount, PurchaseCount manualPurchaseCount) {
+        System.out.println(
+                RESULT_BOUGHT_1 + manualPurchaseCount.getPurchaseCount() + RESULT_BOUGHT_2
+                        + autoPurchaseCount.getPurchaseCount() + RESULT_BOUGHT_3);
```
`System.out.printf` 또는 `String.format`을 사용하면 더 가독성있는 문자열을 만들수 있을것 같아요!

geunhwanlee commented on src/main/java/lotto/Validate.java:
```diff
@@ -96,9 +85,16 @@ private static int isSixNumbers(String input) {
 
-    static void validateBonusNumberDuplicate(int input, List<LottoNumber> winningNumber) {
-        if (winningNumber.contains(new LottoNumber(input))) {
+    static void validateBonusNumberDuplicate(String input, LottoNumbers winningNumber) {
```
유효성검사를 하기 위해 객체를 생성하는 등
점차 유효성검사가 객체에 영향을 받게 되는거같네요.

그러므로 유효성검사를 적절한 객체에게 위임하는건 어떨까요?🤔

geunhwanlee reviewed (changes requested):
5단계 미션 잘해주셨어요!

굉장히 많은 부분을 개선하려고 노력한 부분이 보기 좋았습니다.

다만 몇가지 아쉬운 부분이 있어서 코멘트 드렸어요 😭

고민해보시고 수정해서 재 리뷰요청 주세요🙏
그럼 화이팅하세요!🔥

seogineer committed:
docs: 일급 컬렉션 Amount 추가 및 변경

seogineer committed:
refactor: TotalMoney, Balance를 Amount로 통합

seogineer committed:
docs: 유효성 검사를 할 때, Optional 사용하는 코드 제거

seogineer committed:
refactor: 유효성 검사를 할 때, Optional 사용하는 코드 제거

seogineer committed:
style: 메소드 기능에 대한 이름이 더 명확하게 변경

seogineer committed:
refactor: TotalMoney, Balance를 Amount로 통합

seogineer committed:
docs: PurchaseQuantity -> Quantity 클래스 이름 변경 내용 추가

seogineer committed:
refactor: 적절한 객체로 유효성 검사 위임 수정 외 2건

- 적절한 객체로 유효성 검사 위임
- PurchaseQuantity -> Quantity로 클래스 이름 변경
- Validate 메소드 이름으로 기능을 추측하기 쉽게 변경

seogineer committed:
docs: 변경 내용 반영

seogineer committed:
refactor: 수동 로또 번호 입력 기능 위임 및 LottoNumbers 반환

- LottoGame 수동 로또 번호 입력 기능을 InputView로 위임
- autoGenerateNumbers(), manualGenerateNumbers()의 반환 타입 Lit<Integer> -> LottoNumbers로 변경

seogineer committed:
docs: 클래스 이름 변경 내용 추가

seogineer committed:
style: 클래스 이름 변경

- PurchaseLottoNumbers -> Lottos
- LottoNumbers -> Lotto

seogineer committed:
refactor: printf로 출력 문구 수정

seogineer committed:
refactor: InputView에서 담당하던 보너스 번호 중복 유효성 검사를 Lotto에 위임

seogineer committed:
docs: printf로 코드 변경에 따른 출력 문구에 대한 내용 수정

seogineer committed:
refactor: printf로 출력 코드 변경 및 출력문 상수 이름 변경

geunhwanlee commented on src/main/java/lotto/LottoNumberGenerator.java:
```diff
@@ -20,22 +18,16 @@ public class LottoNumberGenerator {
-        Validate.validateLottoNumberDuplicate(input);
-        return spliter.splitToList(input);
+    public Lotto manualGenerateNumbers() {
+        return inputView.inputManualLottoNumber();
```
view에 대한 의존성을 controller가 아닌 도메인 객체가 가지고있네요..😭
저는 이런 케이스에서는 Lotto를 리턴하는 함수를 파라미터로 넘겨받아서 처리했었어요! 😀
```suggestion
    public Lotto manualGenerateNumbers(Supplier<Lotto> supplier) {
        return supplier.get();
```
이런식이 되겠네요.

혹은 LottoNumberGenerator의 상위 인터페이스를 두고 ([함수형 인터페이스](https://bcp0109.tistory.com/313))

> 1. `자동으로 Lotto를 만드는 구현클래스`를 주입
> 2. controller에서 `view를 통해 수동으로 번호를 입력받아 Lotto를 리턴하는 함수`를 주입

위 방식으로 사용할수도 있겠네요!

혹은 controller에서 수동 로또 `수량`과 `번호`를 입력받은 후 파라미터로 넘겨주는 방법도 있겠네요!

방법은 여러가지가 있을수있고 정답은 없으니 참고정도만 부탁드려요. 😀

geunhwanlee reviewed (changes requested):
리뷰사항 반영 너무 잘해주셨어요! 👍🏻
너무 고생 많으셨습니다. 😭

다만 한가지,
LottoNumberGenerator에서 InputView를 의존하고있는 부분만 개선하면 될것같아요!

한번 시도해보시고 어려운점 있으면 언제든 연락주세요!🙏
거의 다 끝나가네요. 마지막까지 화이팅입니다!🔥

seogineer committed:
style: 클래스 이름 변경에 따라 test 클래스 이름도 변경

seogineer committed:
fix: rank가 제대로 반환되지 않는 문제 수정

예를 들어, 3개를 맞추고 보너스 번호까지 맞춘 경우, FIFTH가 반환되어야 하지만 filter 조건에 matchBonus 조건 때문에 FIFTH가 반환되지 않고 있었음

seogineer committed:
refactor: controller에서 InputView의 inputManualLottoNumber()를 주입 받아 LottoNumberGenerator의 수동 번호 생성을 구현

geunhwanlee reviewed (approved):
마지막 단계까지 고생 많으셨어요!

충분히 잘해주셔서 머지하도록 할게요.

부족한 리뷰지만 도경님께 도움이 되었으면 좋겠네요. ☺️

남은 미션들도 화이팅하시고 많은걸 얻어가셨으면 좋겠습니다! 🙏
